### ISSUE TITLE
Failed SSL connections leak sockets

### ISSUE THREAD

**Issue by WGH-** [weechat/weechat#358]
Apparently, every time SSL connection is failed, a socket is leaked. My weechat has been running for a week I think, trying to connect to some server and always failing it due to certificate problem, and open socket count just keeps increasing.
```
% sockstat | grep weechat | wc -l
569
```
According to `netstat`, they're all in the closed state.

I'm running weechat on FreeBSD, gnutls is used.

**Comment by flashcode** [comment]
Hi,
What's your WeeChat version?
And when you say "SSL connection fails", for which reason? Do you have examples of errors displayed in server buffer?

**Comment by WGH-** [comment]
Version 1.1.1

```
irc: reconnecting to server in 30 minutes
irc: reconnecting to server...
irc: connecting to server irc6.rizon.net/9999 (SSL)...
gnutls: connected using 2048-bit Diffie-Hellman shared secret exchange
gnutls: receiving 1 certificate
 - certificate[1] info:
   - subject `C=RZ,ST=Nowhere,L=Rizon,CN=irc.rizon.net,O=Rizon IRC Network,OU=irc.broke-it.com', issuer `C=RZ,ST=Nowhere,L=Rizon,CN=irc.rizon.net,O=Rizon IRC Network,OU=irc.broke-it.com', RSA key 2048 bits, signed using RSA-SHA1, activated `2013-07-24 21:52:39 UTC', expires `2016-07-23 21:52:39 UTC', SHA-1 fingerprint `150159fe4c0916c13424a312bc2832232922d9e1'
gnutls: the hostname in the certificate does NOT match "irc6.rizon.net"
gnutls: peer's certificate is NOT trusted
gnutls: peer's certificate issuer is unknown
irc: TLS handshake failed
irc: error: Error in the certificate.
irc: reconnecting to server in 30 minutes
```

In case it's helpful, this's how I connect to that server:
```
rizon.addresses = "irc6.rizon.net/9999"
rizon.proxy
rizon.ipv6
rizon.ssl = on
rizon.ssl_cert
rizon.ssl_priorities
rizon.ssl_dhkey_size
rizon.ssl_fingerprint = ""
rizon.ssl_verify
rizon.password
rizon.capabilities
rizon.sasl_mechanism
rizon.sasl_username
rizon.sasl_password
rizon.sasl_timeout
rizon.sasl_fail
rizon.autoconnect = on
rizon.autoreconnect
rizon.autoreconnect_delay
rizon.nicks
rizon.username
rizon.realname
rizon.local_hostname
rizon.command = "/quote privmsg NickServ :IDENTIFY :) :) :)"
rizon.command_delay
rizon.autojoin
rizon.autorejoin
rizon.autorejoin_delay
rizon.connection_timeout
rizon.anti_flood_prio_high
rizon.anti_flood_prio_low
rizon.away_check
rizon.away_check_max_nicks
rizon.default
```

**Comment by Shawn-Smith** [comment]
I wonder if this is in anyway related to #311

**Comment by flashcode** [comment]
@Shawn-Smith: you're right, both problems may be related, I'll check that.

**Comment by flashcode** [comment]
I just tried on Debian and FreeBSD 10.1, with the same server as you: I have the same connection error, and no leak on sockets.
Are you sure the socket leak is caused by this failing connection, or could it be something else?

**Comment by WGH-** [comment]
> Are you sure the socket leak is caused by this failing connection, or could it be something else?

Yes. As you can see, all those everincreasing IPv6 addresses belong to irc6.rizon.net.

```
chireiden% host  irc6.rizon.net
irc6.rizon.net has IPv6 address 2001:1ad8:2002:2::babe
irc6.rizon.net has IPv6 address 2a00:1a28:1100:2::6667
irc6.rizon.net has IPv6 address 2607:2200:0:2351::1783:5213
irc6.rizon.net has IPv6 address 2607:f128:22:3133:312:18:313:11b2
chireiden% sockstat | grep weechat
wgh      dtach      49166 3  stream -> /home/wgh/.weechat/dtach.sock
wgh      weechat    36248 6  tcp6   2a01:4f8:d13:1281:0:2205:1998:849:23034 2001:6b0:e:2a18:5054:ff:fe01:8119:7000
wgh      weechat    36248 7  tcp6   2a01:4f8:d13:1281:0:2205:1998:849:31498 2a00:1a28:1100:2::6667:9999
wgh      weechat    36248 8  tcp6   2a01:4f8:d13:1281:0:2205:1998:849:43888 2001:41d0:52:600::4b4:6667
wgh      weechat    36248 9  tcp4   78.47.122.226:31497   212.53.167.118:6667
wgh      weechat    36248 10 tcp4   78.47.122.226:31495   85.17.239.43:6667
wgh      weechat    36248 11 tcp6   2a01:4f8:d13:1281:0:2205:1998:849:31499 2001:1ad8:2002:2::babe:9999
wgh      weechat    36248 12 tcp6   2a01:4f8:d13:1281:0:2205:1998:849:31496 2001:16d8:aaaa:2::1337:6667
wgh      weechat    36248 13 tcp6   2a01:4f8:d13:1281:0:2205:1998:849:21006 2607:f128:22:3133:312:18:313:11b2:9999
wgh      weechat    36248 14 tcp6   2a01:4f8:d13:1281:0:2205:1998:849:41204 2001:1ad8:2002:2::babe:9999
wgh      weechat    36248 15 tcp6   2a01:4f8:d13:1281:0:2205:1998:849:44910 2a00:1a28:1100:2::6667:9999
wgh      weechat    36248 16 tcp6   2a01:4f8:d13:1281:0:2205:1998:849:13552 2a00:1a28:1100:2::6667:9999
wgh      weechat    36248 17 tcp6   2a01:4f8:d13:1281:0:2205:1998:849:30314 2607:2200:0:2351::1783:5213:9999
wgh      weechat    36248 18 tcp6   2a01:4f8:d13:1281:0:2205:1998:849:34121 2607:2200:0:2351::1783:5213:9999
wgh      weechat    36248 19 tcp6   2a01:4f8:d13:1281:0:2205:1998:849:23801 2607:2200:0:2351::1783:5213:9999
wgh      weechat    36248 20 tcp6   2a01:4f8:d13:1281:0:2205:1998:849:16851 2607:f128:22:3133:312:18:313:11b2:9999
wgh      weechat    36248 21 tcp6   2a01:4f8:d13:1281:0:2205:1998:849:48008 2607:2200:0:2351::1783:5213:9999
wgh      weechat    36248 22 tcp6   2a01:4f8:d13:1281:0:2205:1998:849:44279 2607:2200:0:2351::1783:5213:9999
wgh      weechat    36248 23 tcp6   2a01:4f8:d13:1281:0:2205:1998:849:50667 2607:2200:0:2351::1783:5213:9999
wgh      weechat    36248 24 tcp6   2a01:4f8:d13:1281:0:2205:1998:849:25486 2a00:1a28:1100:2::6667:9999
wgh      weechat    36248 25 tcp6   2a01:4f8:d13:1281:0:2205:1998:849:48223 2001:1ad8:2002:2::babe:9999
wgh      weechat    36248 26 tcp6   2a01:4f8:d13:1281:0:2205:1998:849:21806 2607:2200:0:2351::1783:5213:9999
wgh      weechat    36248 27 tcp6   2a01:4f8:d13:1281:0:2205:1998:849:61423 2a00:1a28:1100:2::6667:9999
wgh      dtach      36247 3  stream /home/wgh/.weechat/dtach.sock
wgh      dtach      36247 5  stream /home/wgh/.weechat/dtach.sock
```

**Comment by flashcode** [comment]
@WGH- : some news about the problem? Do you still have the problem? If yes, on which FreeBSD version?

**Comment by WGH-** [comment]
Nothing's changed. 
FreeBSD 10.1-RELEASE-p19

**Comment by tony** [comment]
FreeBSD 10.2

1.1.1

**Comment by kg6hum** [comment]
I am seeing this same issue.  It looks a few month of running to start getting irc: out of memory errors, but now I can easily reproduce it on an Ubuntu 14.04 system.  I am using weechat-devel-curses from the repository.

Here is the version information: WeeChat 1.6-dev (git: v1.5-99-gb7e9913) [compiled on Jul 17 2016 01:18:45]

I run weechat-curses without a connection to any servers.
I then run: /connect chat.freenode.net/6667 -ssl
That tries to connect to a non-ssl port using ssl, which of course fails.  If I leave this running for a few hours, it will try to connect every 10 minutes.  After many hours, I can run sudo lsof -p pidofweechat to get a list of all connections and files it has open.  I will see something like this:
weechat-c 14450 kg6hum   16u  sock                0,7      0t0  200313 can't identify protocol
weechat-c 14450 kg6hum   17u  sock                0,7      0t0  201972 can't identify protocol
weechat-c 14450 kg6hum   18u  sock                0,7      0t0  202035 can't identify protocol
weechat-c 14450 kg6hum   19u  sock                0,7      0t0  197985 can't identify protocol
weechat-c 14450 kg6hum   20u  sock                0,7      0t0  206921 can't identify protocol
weechat-c 14450 kg6hum   21u  sock                0,7      0t0  198341 can't identify protocol
weechat-c 14450 kg6hum   22u  sock                0,7      0t0  211088 can't identify protocol
weechat-c 14450 kg6hum   23u  sock                0,7      0t0  211747 can't identify protocol
weechat-c 14450 kg6hum   24u  sock                0,7      0t0  231288 can't identify protocol
weechat-c 14450 kg6hum   25u  sock                0,7      0t0  263275 can't identify protocol
weechat-c 14450 kg6hum   26u  IPv4             261991      0t0     TCP 192.168.0.104:52298->198.252.144.2:ircd (CLOSE_WAIT)
weechat-c 14450 kg6hum   27u  IPv4             274513      0t0     TCP 192.168.0.104:52416->198.252.144.2:ircd (CLOSE_WAIT)
weechat-c 14450 kg6hum   28u  IPv4             316537      0t0     TCP 192.168.0.104:52491->198.252.144.2:ircd (CLOSE_WAIT)
weechat-c 14450 kg6hum   29u  IPv4             274823      0t0     TCP 192.168.0.104:52563->198.252.144.2:ircd (CLOSE_WAIT)
weechat-c 14450 kg6hum   30u  IPv4             275897      0t0     TCP 192.168.0.104:52640->198.252.144.2:ircd (CLOSE_WAIT)
weechat-c 14450 kg6hum   31u  IPv4             337339      0t0     TCP 192.168.0.104:53028->198.252.144.2:ircd (CLOSE_WAIT)
weechat-c 14450 kg6hum   32u  IPv4             364608      0t0     TCP 192.168.0.104:53851->198.252.144.2:ircd (CLOSE_WAIT)
weechat-c 14450 kg6hum   33u  IPv4             386197      0t0     TCP 192.168.0.104:54180->198.252.144.2:ircd (CLOSE_WAIT)
weechat-c 14450 kg6hum   34u  IPv4             397580      0t0     TCP 192.168.0.104:54438->198.252.144.2:ircd (CLOSE_WAIT)
weechat-c 14450 kg6hum   35u  IPv4             411888      0t0     TCP 192.168.0.104:54526->198.252.144.2:ircd (CLOSE_WAIT)
weechat-c 14450 kg6hum   36u  IPv4             426614      0t0     TCP 192.168.0.104:54933->198.252.144.2:ircd (CLOSE_WAIT)
weechat-c 14450 kg6hum   37u  IPv4             448566      0t0     TCP 192.168.0.104:55408->198.252.144.2:ircd (ESTABLISHED)
weechat-c 14450 kg6hum   38u  IPv4             473656      0t0     TCP 192.168.0.104:56339->198.252.144.2:ircd (CLOSE_WAIT)

The older sockets show up as "can't identify protocol" and the more recent ones show up in a CLOSE_WAIT state.  I do not see this issue with non-ssl connections, so it seems to be a bug in the cleanup for ssl connections only.  I can do additional testing if needed.

**Comment by kg6hum** [comment]
Can anyone having this issue test the pull request I created?  It has fixed the issue for me.

**Comment by flashcode** [comment]
Closed by #801
